src/smart_chat_bot/src/smart_chat_bot/main.py

```diff
@@ -0,0 +1,208 @@
+logger = logging.getLogger("smart_chat_bot")
+
+CHAT_PROVIDER = os.environ.get("CHAT_PROVIDER", "discord")
+CHANNEL_IDS: list[str] = [
+    channel.strip()
+    for channel in os.environ.get("CHAT_CHANNEL_IDS", "").split(",")
+    if channel.strip()
+]
+if not CHANNEL_IDS:
+    missing_channels_msg = (
+        "CHAT_CHANNEL_IDS must list at least one channel id (comma-separated)."
+    )
+    raise RuntimeError(missing_channels_msg)
+
+POLL_INTERVAL_SECONDS = int(os.environ.get("POLL_INTERVAL_SECONDS", "8"))
+MAX_MESSAGES_PER_POLL = int(os.environ.get("MAX_MESSAGES_PER_POLL", "5"))
+BOT_USER_ID = os.environ.get("BOT_USER_ID")  # optional: skip echoing own messages
+SYSTEM_PROMPT = os.environ.get(
+    "SMART_BOT_SYSTEM_PROMPT",
+    "You are a concise, helpful assistant.",
+)
+
+app = FastAPI(
+    title="Smart Chat Bot API",
+    description="FastAPI service that polls chat provider, runs AI, and posts replies.",
+    version="1.0.0",
+)
+
+
+# ---------------------------
+# Helpers
+# ---------------------------
+
+T = TypeVar("T")
+
+
+async def _to_thread(func: Callable[..., T], *args: object, **kwargs: object) -> T:
+    return await asyncio.to_thread(func, *args, **kwargs)
+
+
+def make_chat_client(provider: str, **cfg: str | None) -> ChatInterface:
+    """Return a ChatInterface implementation based on the selected provider."""
+    if provider == "discord":
+        token = cfg.get("access_token") or os.environ.get("DISCORD_BOT_TOKEN")
+        token_type = cfg.get("token_type") or os.environ.get("DISCORD_TOKEN_TYPE", "Bot")
+        if token:
+            return DiscordClient(access_token=token, token_type=token_type)
+        return chat_client_api.get_client()
+
+    msg = f"Unknown chat provider: {provider}"
+    raise ValueError(msg)
+
+
+async def fetch_recent_messages(client: ChatInterface, channel_id: str, limit: int) -> list[Message]:
+    """Fetch recent messages (synchronously) in a thread."""
+    return await _to_thread(client.get_messages, channel_id, limit=limit)
+
+
+async def send_message(client: ChatInterface, channel_id: str, content: str) -> bool:
+    """Send a message via the chat client in a thread."""
+    return await _to_thread(client.send_message, channel_id, content)
+
+
+async def generate_reply(ai: AIInterface, user_input: str) -> str:
+    """Generate AI reply in a thread."""
+    result = await _to_thread(ai.generate_response, user_input, SYSTEM_PROMPT, None)
+    return result if isinstance(result, str) else str(result)
+
+
+def _iter_new_messages(messages: Iterable[Message], last_seen_id: str | None) -> Iterable[Message]:
+    """Yield messages that are newer than last_seen_id (messages ordered newest->oldest)."""
+    for msg in messages:
+        if last_seen_id and msg.id == last_seen_id:
+            break
+        yield msg
+
+
+def is_bot_message(msg: Message) -> bool:
+    """Return True if the message author is marked as a bot."""
+    raw = getattr(msg, "_raw_data", None)
```

Comment:
This only works for a specific implementation... Sure, `None` default, but still expecting more than abstract...